abci/types/types.go

```diff
@@ -167,3 +170,125 @@ func RespondVerifyVoteExtension(ok bool) ResponseVerifyVoteExtension {
 		Result: result,
 	}
 }
+
+func MustHashResults(r []*ExecTxResult) []byte {
+	return merkle.HashFromByteSlices(mustResultsToByteSlices(r))
+}
+
+// ProveResult returns a merkle proof of one result from the set
+func MustProveResult(r []*ExecTxResult, i int) merkle.Proof {
+	_, proofs := merkle.ProofsFromByteSlices(mustResultsToByteSlices(r))
+	return *proofs[i]
+}
+
+// deterministicExecTxResult strips non-deterministic fields from
+// ResponseDeliverTx and returns another ResponseDeliverTx.
+func deterministicExecTxResult(response *ExecTxResult) *ExecTxResult {
+	return &ExecTxResult{
+		Code:      response.Code,
+		Data:      response.Data,
+		GasWanted: response.GasWanted,
+		GasUsed:   response.GasUsed,
+	}
+}
+
+func mustResultsToByteSlices(r []*ExecTxResult) [][]byte {
+	s := make([][]byte, len(r))
+	for i, e := range r {
+		d := deterministicExecTxResult(e)
+		b, err := d.Marshal()
+		if err != nil {
+			panic(err)
+		}
+		s[i] = b
+	}
+	return s
+}
+
+func (tr *TxRecord) IsIncluded() bool {
+	return tr.Action == TxRecord_ADDED || tr.Action == TxRecord_UNMODIFIED
+}
+
+// IncludedTxs returns all of the TxRecords that are marked for inclusion in the
+// proposed block.
+func (rpp *ResponsePrepareProposal) IncludedTxs() []*TxRecord {
+	trs := []*TxRecord{}
+	for _, tr := range rpp.TxRecords {
+		if tr.IsIncluded() {
+			trs = append(trs, tr)
+		}
+	}
+	return trs
+}
+
+// RemovedTxs returns all of the TxRecords that are marked for removal from the
+// mempool.
+func (rpp *ResponsePrepareProposal) RemovedTxs() []*TxRecord {
+	trs := []*TxRecord{}
+	for _, tr := range rpp.TxRecords {
+		if tr.Action == TxRecord_REMOVED {
+			trs = append(trs, tr)
+		}
+	}
+	return trs
+}
+
+// AddedTxs returns all of the TxRecords that are marked as added to the proposal.
+func (rpp *ResponsePrepareProposal) AddedTxs() []*TxRecord {
+	trs := []*TxRecord{}
+	for _, tr := range rpp.TxRecords {
+		if tr.Action == TxRecord_ADDED {
+			trs = append(trs, tr)
+		}
+	}
+	return trs
+}
+
+// Validate checks that the fields of the ResponsePrepareProposal are properly
+// constructed. Validate returns an error if any of the validation checks fail.
+func (rpp *ResponsePrepareProposal) Validate(maxSizeBytes int64, otxs [][]byte) error {
+
+	if !rpp.ModifiedTx {
+		// This method currently only checks the validity of the TxRecords field.
+		// If ModifiedTx is false, then we can ignore the validity of the TxRecords field.
+		//
+		// TODO: When implementing VoteExensions, AppSignedUpdates may be modified by the application
+		// and this method should be updated to validate the AppSignedUpdates.
+		return nil
+	}
+
+	// TODO: this feels like a large amount allocated data. We move all the Txs into strings
+	// in the map. The map will be as large as the original byte slice.
+	// Is there a key we can use?
```

Comment:
We could map integer indices into the original slice-of-slices instead of copies of the data. We don't permute the input slice.

Alternatively: We could hash them. Any strong hash should be fine—we're not persisting these values, it's just for duplicate detection.

Alternatively: Map a weak hash to slices of candidates, and probe the candidates when the weak hash hits. I don't think we need to optimize here yet—these are all transient, and copying is almost certainly cheaper than hashing.

Reply:
> We could map integer indices into the original slice-of-slices instead of copies of the data. We don't permute the input slice.

Would you mind elaborating on what this would look like? I'm not totally sure I understand. There sort of already is a mapping of integer to []byte{} via their position in the slice.

Reply:
Wouldn't this involve defining `otxSet := make(map[int]struct{}, len(otxs))` whose keys would be a numeric hash of each original tx? Does Go have an equivalent of Java's `hashCode()` for this purpose? Cryptographic hashes would be computationally expensive in comparison, so you'd use less memory at the cost of slower validation (especially for large batches of txs).

Reply:
There is no built-in hash code. But that's more or less what I meant by a weak hash, say a CRC or Adler checksum. That's enough to quickly eliminate true negatives and cheaper to compute than a cryptographic hash. 

The breakpoint in size where I would expect strong hashing to help is probably quite high for the reasons @thanethomson suggests.

Re: mapping. I mean that one can eliminate any copies of data by keeping say a `map[uint32]int` where the keys are CRC32s and the values are the indices of the corresponding slice in the input.